controller/api/destination/watcher/k8s.go

```diff
@@ -41,3 +44,17 @@ func invalidService(authority string) InvalidService {
 func (i ID) String() string {
 	return fmt.Sprintf("%s/%s", i.Namespace, i.Name)
 }
+
+// Used by the delete handlers to make sure we are getting the
+// correct type out of the DeltaFIFO
+func extractDeletedObject(obj interface{}, expectedType reflect.Type) (interface{}, error) {
+	switch object := obj.(type) {
+	case cache.DeletedFinalStateUnknown:
```

Comment:
It seems to be a very common pattern in the K8s code to validate the type of `object.Obj` too. I'd recommend doing the same thing. In that case, it might be easier to just not use a helper function. See

* Replicaset : https://github.com/kubernetes/kubernetes/blob/0b2636a7e705bdf7c911e12d2e8846b9e6f88704/pkg/controller/replicaset/replica_set.go#L330-L342
* Endpoint: https://github.com/kubernetes/kubernetes/blob/0b2636a7e705bdf7c911e12d2e8846b9e6f88704/pkg/controller/util/endpoint/controller_utils.go#L219-L235
* Deployment: https://github.com/kubernetes/kubernetes/blob/0b2636a7e705bdf7c911e12d2e8846b9e6f88704/pkg/controller/deployment/deployment_controller.go#L339-L356
 * Statefulset: https://github.com/kubernetes/kubernetes/blob/0b2636a7e705bdf7c911e12d2e8846b9e6f88704/pkg/controller/statefulset/stateful_set.go#L243-L259

WDYT?

Reply:
Good point. Should have check how they do it. I am not a programming language expert but am I the only one finding that doing so much type checking in a typed programming language is a shame :(

Reply:
Yeah, generally a function parameter or struct field of `interface` type is useless, and it's a smell. If only there the language supports generic data types...